prov/gni/src/gnix_eq.c

```diff
@@ -170,22 +241,195 @@ static int gnix_eq_close(struct fid *fid)
+		goto cleanup;
+	}
+
+	*event = ev->type;
+	memcpy(buf, ev->buf, len);
+
+	/*
+	 * Remove if FI_PEEK isn't specified.
+	 */
+	if (!(flags & FI_PEEK)) {
+		ev = container_of(slist_remove_head(&gnix_eq->ev_queue),
+				  struct gnix_event, entry);
+		free(ev->buf);
+		free(ev);
+	}
+
+	fastlock_release(&gnix_eq->lock);
+
+	return ret;
+
+cleanup:
+	fastlock_release(&gnix_eq->lock);
+err:
+	return ret;
 }
 
 static int gnix_eq_readerr(struct fid_eq *eq, struct fi_eq_err_entry *buf,
 			   size_t len, uint64_t flags)
 {
-	return -FI_ENOSYS;
+	struct gnix_fid_eq *gnix_eq = NULL;
+	struct gnix_event *ev = NULL;
+	int ret = len;
+
+	gnix_eq = container_of(eq, struct gnix_fid_eq, eq_fid);
+	if (!gnix_eq) {
+		ret = -FI_EINVAL;
+		goto err;
+	}
+
+	fastlock_acquire(&gnix_eq->lock);
+
+	/*
+	 * If there is no error, then signal to try again.
+	 */
+	if (slist_empty(&gnix_eq->err_queue)) {
+		ret = -FI_EAGAIN;
+		goto cleanup;
+	}
+
+	ev = container_of(gnix_eq->err_queue.head, struct gnix_event, entry);
+	if (!event) {
+		ret = -FI_EINVAL;
+		goto cleanup;
+	}
+
+	/*
+	 * Not a good idea to truncate an event that will be removed from the
+	 * queue.
+	 */
+	if (len < ev->len) {
+		ret = -FI_ETOOSMALL;
+		goto cleanup;
+	}
+
+	memcpy(buf, ev->buf, len);
+
+	/*
+	 * Remove if FI_PEEK isn't specified.
+	 */
+	if (!(flags & FI_PEEK)) {
+		ev = container_of(slist_remove_head(&gnix_eq->err_queue),
+				  struct gnix_event, entry);
+		free(ev->buf);
+		free(ev);
+	}
+
+	fastlock_release(&gnix_eq->lock);
+
+	return ret;
+
+cleanup:
+	fastlock_release(&gnix_eq->lock);
+err:
+	return ret;
 }
 
 static int gnix_eq_write(struct fid_eq *eq, uint32_t event,
 			 const void *buf, size_t len, uint64_t flags)
 {
-	return -FI_ENOSYS;
+	struct gnix_fid_eq *gnix_eq = NULL;
+	struct gnix_event *ev_entry = NULL;
+	int ret = len;
+	void *ev_buf = NULL;
+
+	/*
+	 * Allocate memory for the event entry that will be stored inside of the
+	 * event queue. Memory for the given buffer containing the fi_eq_entry
+	 * must be allocated separately.
+	 */
+	ev_entry = calloc(1, sizeof(*ev_entry));
+	if (!ev_entry) {
+		ret = -FI_ENOMEM;
+		goto err;
+	}
+
+	/*
+	 * The buf entry is meant to contain a type of fi_eq_entry, but this is
+	 * only a header and it could be larger. Allocate size len which is
+	 * assumed to already be sizeof(fi_eq_entry) + sizeof(extra_data).
+	 */
+	ev_buf = calloc(1, len);
+	if (!ev_buf) {
+		ret = -FI_ENOMEM;
+		goto cleanup_ev_entry;
+	}
+
+	gnix_eq = container_of(eq, struct gnix_fid_eq, eq_fid);
+	if (!gnix_fid_eq) {
+		ret = -FI_EINVAL;
+		goto cleanup_ev_buf;
+	}
+
+	memcpy(ev_buf, buf, len);
+	ev_entry->buf = ev_buf;
+
+	ev_entry->type = event;
+	ev_entry->len = len;
+	ev_entry->flags = flags;
+
+	/*
+	 * TODO:
+	 * - Handle fastlock_acquire failure case.
+	 */
+	fastlock_acquire(&gnix_eq->lock);
+
+	/*
+	 * Insert generated entry into the event queue.
+	 */
+	slist_insert_tail(&ev_entry->entry, &gnix_eq->ev_queue);
+
+	fastlock_release(&gnix_eq->lock);
```

Comment:
Food for thought: Might consider a wrapper for the set of these 3 operations (acquire, insert, release) with the error check if it become more common.